compiler/rustc_middle/src/ty/relate.rs

```diff
@@ -59,8 +59,10 @@ pub trait TypeRelation<'tcx>: Sized {
             item_def_id, a_subst, b_subst
         );
 
-        let opt_variances = self.tcx().variances_of(item_def_id);
-        relate_substs(self, Some(opt_variances), a_subst, b_subst)
+        let tcx = self.tcx();
+        let opt_variances = tcx.variances_of(item_def_id);
+        let get_ty = || tcx.type_of(item_def_id).subst(tcx, a_subst);
+        relate_substs(self, Some((&get_ty, opt_variances)), a_subst, b_subst)
```

Comment:
We don't really need to take a `dyn Fn`. Could we get away with passing `item_def_id` and have `relate_substs` call `type_of` and substitute it?